### PR TITLE
Fix fp16 support for cuda shared array

### PR DESCRIPTION
**Pull request by testhound**
This pull request fixes an issue identified in the fp16 master issue: 

https://github.com/numba/numba/issues/4402

where the new float16 data type was not accepted as a data type for 'cuda.shared.array'

### CODE REVIEW

**Comment by testhound** [comment]
gpuci run tests

**Comment by gmarkall** [comment]
gpuci run tests

File: numba/cuda/cudaimpl.py [hunk 1]
```diff
@@ -992,6 +992,7 @@ def _generic_array(context, builder, shape, dtype, symbol_name, addrspace,
     other_supported_type = (
         isinstance(dtype, (types.Record, types.Boolean))
         or isinstance(data_model, models.StructModel)
+        or dtype == types.float16
```
Comment (gmarkall) [inline numba/cuda/cudaimpl.py:995]: What do you think about adding `types.float16` to `types.number_domain` instead of making it a special case here?

**Review by gmarkall — CHANGES_REQUESTED** [review]
I did a quick search of the codebase for `number_domain` to look for other areas where we might have issues with `float16` support. One is with built-in casts - for example,

```python
from numba import cuda
from numba.types import float16, float32

def f(x):
    return float(x)

print("Compiling for float32...")
cuda.compile_ptx(f, (float32,), device=True)
print("Compiling for float16...")
cuda.compile_ptx(f, (float16,), device=True)
```

gives:

```
$ python cast.py 
Compiling for float32...
Compiling for float16...

...

numba.core.errors.TypingError: No implementation of function Function(<class 'float'>) found for signature:
 
 >>> float(float16)
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'Float.generic': File: numba/core/typing/builtins.py: Line 968.
    With argument(s): '(float16)':
   Rejected as the implementation raised a specific error:
     NumbaTypeError: float() only support for numbers
       Traceback (most recent call last):
         File "/home/gmarkall/numbadev/numba/numba/core/types/functions.py", line 308, in get_call_type
           sig = temp.apply(nolitargs, nolitkws)
         File "/home/gmarkall/numbadev/numba/numba/core/typing/templates.py", line 351, in apply
           sig = generic(args, kws)
         File "/home/gmarkall/numbadev/numba/numba/core/typing/builtins.py", line 974, in generic
           raise errors.NumbaTypeError("float() only support for numbers")
       numba.core.errors.NumbaTypeError: float() only support for numbers
  raised from /home/gmarkall/numbadev/numba/numba/core/typing/builtins.py:974
```

and the other is with using Numba types for shared memory - e.g.

```python
from numba import cuda, types

print('Definining for types.float32...')

@cuda.jit('void()')
def f_32():
    cuda.shared.array(1, dtype=types.float32)

print('Definining for types.float16...')

@cuda.jit('void()')
def f_16():
    cuda.shared.array(1, dtype=types.float16)

```

gives:

```
$ python shared.py 
Definining for types.float32...
Definining for types.float16...
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/issues/7972/shared.py", line 16, in <module>
    def f_16():

...

  File "/home/gmarkall/numbadev/numba/numba/cuda/cudaimpl.py", line 160, in cuda_shared_array_integer
    return _generic_array(context, builder, shape=(length,), dtype=dtype,
  File "/home/gmarkall/numbadev/numba/numba/cuda/cudaimpl.py", line 1013, in _generic_array
    raise TypeError("unsupported type: %s" % dtype)
TypeError: unsupported type: float16
```

There's a couple of other things I noticed:

- Implementing support for a `complex32` will need a similar audit for uses of `number_domain` (or eventual addition of `float16` to `number_domain`)
- Kwargs support for CUDA (e.g. as in #7871) will need some consideration because `typeof_init` in `_typeof.c` is not called for `float16`.

Moving forward:

- WRT to the first issue, `cast.py` above - could you either fix that in this PR or open an issue describing the situation please?
- WRT to the second issue, `shared.py` above - could you resolve that in this PR please?

**Comment by testhound** [comment]
gpuci run tests

**Comment by testhound** [comment]
> I did a quick search of the codebase for `number_domain` to look for other areas where we might have issues with `float16` support. One is with built-in casts - for example,
> 
> ```python
> from numba import cuda
> from numba.types import float16, float32
> 
> def f(x):
>     return float(x)
> 
> print("Compiling for float32...")
> cuda.compile_ptx(f, (float32,), device=True)
> print("Compiling for float16...")
> cuda.compile_ptx(f, (float16,), device=True)
> ```
> 
> gives:
> 
> ```
> $ python cast.py 
> Compiling for float32...
> Compiling for float16...
> 
> ...
> 
> numba.core.errors.TypingError: No implementation of function Function(<class 'float'>) found for signature:
>  
>  >>> float(float16)
>  
> There are 2 candidate implementations:
>   - Of which 2 did not match due to:
>   Overload in function 'Float.generic': File: numba/core/typing/builtins.py: Line 968.
>     With argument(s): '(float16)':
>    Rejected as the implementation raised a specific error:
>      NumbaTypeError: float() only support for numbers
>        Traceback (most recent call last):
>          File "/home/gmarkall/numbadev/numba/numba/core/types/functions.py", line 308, in get_call_type
>            sig = temp.apply(nolitargs, nolitkws)
>          File "/home/gmarkall/numbadev/numba/numba/core/typing/templates.py", line 351, in apply
>            sig = generic(args, kws)
>          File "/home/gmarkall/numbadev/numba/numba/core/typing/builtins.py", line 974, in generic
>            raise errors.NumbaTypeError("float() only support for numbers")
>        numba.core.errors.NumbaTypeError: float() only support for numbers
>   raised from /home/gmarkall/numbadev/numba/numba/core/typing/builtins.py:974
> ```
> 
> and the other is with using Numba types for shared memory - e.g.
> 
> ```python
> from numba import cuda, types
> 
> print('Definining for types.float32...')
> 
> @cuda.jit('void()')
> def f_32():
>     cuda.shared.array(1, dtype=types.float32)
> 
> print('Definining for types.float16...')
> 
> @cuda.jit('void()')
> def f_16():
>     cuda.shared.array(1, dtype=types.float16)
> ```
> 
> gives:
> 
> ```
> $ python shared.py 
> Definining for types.float32...
> Definining for types.float16...
> Traceback (most recent call last):
>   File "/home/gmarkall/numbadev/issues/7972/shared.py", line 16, in <module>
>     def f_16():
> 
> ...
> 
>   File "/home/gmarkall/numbadev/numba/numba/cuda/cudaimpl.py", line 160, in cuda_shared_array_integer
>     return _generic_array(context, builder, shape=(length,), dtype=dtype,
>   File "/home/gmarkall/numbadev/numba/numba/cuda/cudaimpl.py", line 1013, in _generic_array
>     raise TypeError("unsupported type: %s" % dtype)
> TypeError: unsupported type: float16
> ```
> 
> There's a couple of other things I noticed:
> 
>     * Implementing support for a `complex32` will need a similar audit for uses of `number_domain` (or eventual addition of `float16` to `number_domain`)
> 
>     * Kwargs support for CUDA (e.g. as in #7871) will need some consideration because `typeof_init` in `_typeof.c` is not called for `float16`.
> 
> 
> Moving forward:
> 
>     * WRT to the first issue, `cast.py` above - could you either fix that in this PR or open an issue describing the situation please?
> 
>     * WRT to the second issue, `shared.py` above - could you resolve that in this PR please?

@gmarkall thanks for the review I fixed the first issue (cast) with the most recent change. The second issue was already resolved with my first initial patch.

**Comment by gmarkall** [comment]
> @gmarkall thanks for the review I fixed the first issue (cast) with the most recent change. The second issue was already resolved with my first initial patch.

Many thanks - can you add a test for the second issue (using `types.float16` instead of `np.float16`) please?

**Comment by gmarkall** [comment]
> Many thanks - can you add a test for the second issue (using `types.float16` instead of `np.float16`) please?

(And also for the native cast too)

**Comment by testhound** [comment]
gpuci run tests

File: numba/cuda/tests/cudapy/test_sm.py [hunk 2]
```diff
@@ -42,8 +42,22 @@ def test_issue_1051_shared_size_broken_2d(self):
         self._check_shared_array_size((2, 3), 6)
 
     def test_issue_1051_shared_size_broken_3d(self):
+
         self._check_shared_array_size((2, 3, 4), 24)
 
+    def _check_shared_array_size_fp16(self, shape, expected):
+        @cuda.jit
+        def s(a):
+            arr = cuda.shared.array(shape, dtype=types.float16)
```
Comment (gmarkall) [inline numba/cuda/tests/cudapy/test_sm.py:51]: Do you have a test for the case where `np.float16` is used?

**Review by gmarkall — CHANGES_REQUESTED** [review]
Thanks for the update. I think both cases (`types.float16` and `numpy.float16`) need to be tested. I think a test for the "native" cast still needs adding too?

File: numba/cuda/tests/cudapy/test_sm.py (hunk 2 above)
Comment (testhound) [reply numba/cuda/tests/cudapy/test_sm.py]: @gmarkall test case added.

**Comment by testhound** [comment]
gpuci run tests

**Review by gmarkall — CHANGES_REQUESTED** [review]
Thanks for the updates. There are still a couple of points to resolve:

- There is no test for the native cast (`float(x)` where `x` is a `float16`).
- The changes to fix the native cast turn what was a typing error on the CPU target into a `NotImplementedError` later on in the pipeline. This is a problem because it's OK to present the user with a typing error (indicating that they're trying to do something that's not supported), but it's not OK to show a traceback from somewhere in the internals of Numba once it's past lowering.

An example of the issue this introduces in the CPU target is with:

```python
from numba import njit
import numpy as np

@njit
def f(x):
    return float(x)

print(f(np.float16(2.3)))
```

which gives

```
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<class 'float'>) found for signature:
 
 >>> float(float16)
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'Float.generic': File: numba/core/typing/builtins.py: Line 968.
    With argument(s): '(float16)':
   Rejected as the implementation raised a specific error:
     NumbaTypeError: float() only support for numbers
       Traceback (most recent call last):
         File "/home/gmarkall/numbadev/numba/numba/core/types/functions.py", line 308, in get_call_type
           sig = temp.apply(nolitargs, nolitkws)
         File "/home/gmarkall/numbadev/numba/numba/core/typing/templates.py", line 351, in apply
           sig = generic(args, kws)
         File "/home/gmarkall/numbadev/numba/numba/core/typing/builtins.py", line 974, in generic
           raise errors.NumbaTypeError("float() only support for numbers")
       numba.core.errors.NumbaTypeError: float() only support for numbers
  raised from /home/gmarkall/numbadev/numba/numba/core/typing/builtins.py:974

During: resolving callee type: Function(<class 'float'>)
During: typing of call at /home/gmarkall/numbadev/numba/7972/cpu.py (7)


File "cpu.py", line 7:
def f(x):
    return float(x)
    ^
```

with the main branch, but:

```
raceback (most recent call last):
  File "/home/gmarkall/numbadev/numba/7972/cpu.py", line 10, in <module>
    print(f(np.float16(2.3)))
  File "/home/gmarkall/numbadev/numba/numba/core/dispatcher.py", line 487, in _compile_for_args
    raise e
  File "/home/gmarkall/numbadev/numba/numba/core/dispatcher.py", line 420, in _compile_for_args
    return_val = self.compile(tuple(argtypes))
  File "/home/gmarkall/numbadev/numba/numba/core/dispatcher.py", line 965, in compile
    cres = self._compiler.compile(args, return_type)
  File "/home/gmarkall/numbadev/numba/numba/core/dispatcher.py", line 125, in compile
    status, retval = self._compile_cached(args, return_type)
  File "/home/gmarkall/numbadev/numba/numba/core/dispatcher.py", line 139, in _compile_cached
    retval = self._compile_core(args, return_type)
  File "/home/gmarkall/numbadev/numba/numba/core/dispatcher.py", line 152, in _compile_core
    cres = compiler.compile_extra(self.targetdescr.typing_context,
  File "/home/gmarkall/numbadev/numba/numba/core/compiler.py", line 716, in compile_extra
    return pipeline.compile_extra(func)
  File "/home/gmarkall/numbadev/numba/numba/core/compiler.py", line 452, in compile_extra
    return self._compile_bytecode()
  File "/home/gmarkall/numbadev/numba/numba/core/compiler.py", line 520, in _compile_bytecode
    return self._compile_core()
  File "/home/gmarkall/numbadev/numba/numba/core/compiler.py", line 495, in _compile_core
    raise e
  File "/home/gmarkall/numbadev/numba/numba/core/compiler.py", line 486, in _compile_core
    pm.run(self.state)
  File "/home/gmarkall/numbadev/numba/numba/core/compiler_machinery.py", line 364, in run
    raise e
  File "/home/gmarkall/numbadev/numba/numba/core/compiler_machinery.py", line 356, in run
    self._runPass(idx, pass_inst, state)
  File "/home/gmarkall/numbadev/numba/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/home/gmarkall/numbadev/numba/numba/core/compiler_machinery.py", line 311, in _runPass
    mutated |= check(pss.run_pass, internal_state)
  File "/home/gmarkall/numbadev/numba/numba/core/compiler_machinery.py", line 273, in check
    mangled = func(compiler_state)
  File "/home/gmarkall/numbadev/numba/numba/core/typed_passes.py", line 394, in run_pass
    lower.lower()
  File "/home/gmarkall/numbadev/numba/numba/core/lowering.py", line 197, in lower
    self.lower_normal_function(self.fndesc)
  File "/home/gmarkall/numbadev/numba/numba/core/lowering.py", line 247, in lower_normal_function
    self.setup_function(fndesc)
  File "/home/gmarkall/numbadev/numba/numba/core/lowering.py", line 321, in setup_function
    self.function = self.context.declare_function(self.module, fndesc)
  File "/home/gmarkall/numbadev/numba/numba/core/base.py", line 416, in declare_function
    fnty = self.call_conv.get_function_type(fndesc.restype, fndesc.argtypes)
  File "/home/gmarkall/numbadev/numba/numba/core/callconv.py", line 482, in get_function_type
    arginfo = self._get_arg_packer(argtypes)
  File "/home/gmarkall/numbadev/numba/numba/core/callconv.py", line 159, in _get_arg_packer
    return self.context.get_arg_packer(argtypes)
  File "/home/gmarkall/numbadev/numba/numba/core/base.py", line 323, in get_arg_packer
    return datamodel.ArgPacker(self.data_model_manager, fe_args)
  File "/home/gmarkall/numbadev/numba/numba/core/datamodel/packer.py", line 81, in __init__
    dm = self._dmm.lookup(ty)
  File "/home/gmarkall/numbadev/numba/numba/core/datamodel/manager.py", line 30, in lookup
    model = self._cache[fetype] = handler(self, fetype)
  File "/home/gmarkall/numbadev/numba/numba/core/datamodel/models.py", line 369, in __init__
    raise NotImplementedError(fe_type)
NotImplementedError: float16
```

with this branch.

Given that it would be good to get this in for 0.56RC, which is quite soon, it might be better to break out the fix for the native cast into a separate issue / PR.

**Comment by testhound** [comment]
gpuci run tests

**Comment by testhound** [comment]
@gmarkall changes to builtin.py reversed and new issue created for native cast.

**Comment by testhound** [comment]
@gmarkall changes to builtin.py reversed and new issue created for native cast.

**Comment by gmarkall** [comment]
Thanks for the update. I was just about to approve this when I noticed that the issue also applies to local memory. Although this PR fixes it for the device (because the code path is common) it doesn't fix it for the simulator, because it has separate code paths for `cuda.shared.array` and `cuda.local.array`. This commit fixes it and adds a similar test for local memory: https://github.com/gmarkall/numba/commit/43b7af0e066c1c7da85670ea488129fb7462ba88 - if you could add this commit to your branch then I think this PR will be ready to approve.

**Comment by testhound** [comment]
gpuci run tests

**Review by gmarkall — APPROVED** [review]
Many thanks for the update - I think this is now looking good!

**Comment by gmarkall** [comment]
@esc / @stuartarchibald Could this have a CUDA buildfarm run please?

**Comment by sklam** [comment]
BFID `numba_smoketest_cuda_yaml_138`

**Comment by esc** [comment]
> BFID `numba_smoketest_cuda_yaml_138`

This was all 🟢

**Comment by esc** [comment]
Thanks for the patch!

**Pull request merged**